### ISSUE TITLE
Neural D

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs-benchmark#198]
The idea is to use a neural network, like this:
```
        self.dropout = 0.7
        self.hidden_dim = 128

        # Neural network for next_d function
        self.difficulty_nn = nn.Sequential(
            nn.Linear(3, self.hidden_dim),  # Input: retrievability, difficulty, rating
            nn.Dropout(self.dropout),
            Mish(),
            nn.Linear(self.hidden_dim, self.hidden_dim),
            nn.Dropout(self.dropout),
            Mish(),
            nn.Linear(self.hidden_dim, 1)  # Output: new difficulty
        )
```

To calculate D. Then we can train that new FSRS on 100 users, and then we can plot the outputs of the neural D function and try to find simple functions to approximate them.

Regarding regularization, weight decay with a smaller net is probably better than dropout with a larger net, but I couldn't get either to work, so idk. I tried this:
```
            self.optimizer = torch.optim.Adam([{'params': self.model.w},
                                               {'params': self.model.difficulty_nn.parameters(), 'weight_decay': 1.0}],
                                               lr=lr)
```
And got exactly the same outputs for wd=10 and wd=0.0001.

So anyway, after you train it, you do something like:
```
for R in r_range:
    for D in d_range:
        for grade in [1, 2, 3, 4]:
            new_d = neural_stuff(R, D, G)

```
Regarding plotting, I'm not sure what exactly would be the best. Four 2D graphs (f(D, R) for each grade) would probably be hard to analyze, maybe @1DWalker has a good idea for a visualization.

Btw, if this approach works and we discover a new, better formula for D, we could use it to figure out a short-term memory formula, too!

**Comment by L-M-Sherlock** [comment]
It's very slow to use such a large neural network. And it doesn't perform well even when I increase the `n_epoch` to 10:

```
Model: FSRS-5-dev
Total number of users: 31
Total number of reviews: 485695
Weighted average by reviews:
FSRS-5-dev LogLoss (mean±std): 0.3571±0.1870
FSRS-5-dev RMSE(bins) (mean±std): 0.0722±0.0296
FSRS-5-dev AUC (mean±std): 0.7085±0.0871

Weighted average by log(reviews):
FSRS-5-dev LogLoss (mean±std): 0.3685±0.1756
FSRS-5-dev RMSE(bins) (mean±std): 0.0727±0.0304
FSRS-5-dev AUC (mean±std): 0.7017±0.0851

Weighted average by users:
FSRS-5-dev LogLoss (mean±std): 0.3701±0.1742
FSRS-5-dev RMSE(bins) (mean±std): 0.0729±0.0305
FSRS-5-dev AUC (mean±std): 0.7011±0.0850

Model: FSRS-5
Total number of users: 31
Total number of reviews: 485695
Weighted average by reviews:
FSRS-5 LogLoss (mean±std): 0.3536±0.1862
FSRS-5 RMSE(bins) (mean±std): 0.0695±0.0300
FSRS-5 AUC (mean±std): 0.7200±0.0863

Weighted average by log(reviews):
FSRS-5 LogLoss (mean±std): 0.3653±0.1745
FSRS-5 RMSE(bins) (mean±std): 0.0703±0.0305
FSRS-5 AUC (mean±std): 0.7105±0.0862

Weighted average by users:
FSRS-5 LogLoss (mean±std): 0.3670±0.1731
FSRS-5 RMSE(bins) (mean±std): 0.0707±0.0306
FSRS-5 AUC (mean±std): 0.7095±0.0866
```

**Comment by Expertium** [comment]
Shouldn't using cuda help with speed? Also, I recommend training it on a combined revlog from many users rather than on individual users.
